DSESwift/DataSyncer.swift

```diff
@@ -8,37 +8,398 @@
+        var queryOptions = QueryOptions()
+        queryOptions.existsInRemote = false
+        return try DatabaseHandler.getDataPoints(sensor.id, queryOptions)
     }
     
-    func enablePeriodicSync() {}
-    func disablePeriodicSync(){}
+    private func uploadDataPointsForSensor(dataPoints: Array<DataPoint>, sensor: Sensor) throws {
+        let dataArray = try DataSyncer.getJSONArray(dataPoints, sensorName: sensor.name)
+        try SensorDataProxy.putSensorData(sourceName: DataSyncer.SOURCE, sensorName: sensor.name, data: dataArray, meta: sensor.meta)
+    }
+    
+    private func updateUploadStatusForDataPoints(dataPoints: Array<DataPoint>) throws {
+        for datapoint in dataPoints {
+            datapoint.existsInRemote = true
+            try DatabaseHandler.insertOrUpdateDataPoint(datapoint)
+        }
+    }
+    
+    private func getSensorsInLocal() -> Array<Sensor>{
+        var allSensorsInLocal = Array<Sensor>()
+        let sources = DatabaseHandler.getSources()
+        for source in sources{
+            allSensorsInLocal.appendContentsOf(DatabaseHandler.getSensors(source))
+        }
+        return allSensorsInLocal
+    }
     
-    func downloadSensorProfile() {
-        MockProxy.getSensorProfile()
+    private func insertSensorsIntoLocal(sensors: Array<Sensor>) throws {
+        for sensor in sensors {
+            if !DatabaseHandler.hasSensor(sensor.source, sensorName: sensor.name) {
+                try DatabaseHandler.insertSensor(sensor)
+            } else {
+                try DatabaseHandler.updateSensor(sensor)
+            }
+        }
+    }
+    
+    private func getSensorsFromRemote() throws -> Array<Sensor>{
+        let downloadedArray = try SensorDataProxy.getSensors()
+        return convertAnyObjArrayToSensorArray(downloadedArray.arrayObject)
+    }
+    
+    private func convertAnyObjArrayToSensorArray(inputArray: Array<AnyObject>?) -> Array<Sensor>{
+        var sensors = Array<Sensor>()
+        for anyObj in inputArray! {
+            sensors.append(convertAnyObjectToSensor(anyObj))
+        }
+        return sensors
+    }
+    
+    private func convertAnyObjectToSensor(anyObj: AnyObject) -> Sensor{
+        let sensorDict = anyObj as! Dictionary<String, AnyObject>
+        let sourceName = sensorDict["source_name"] as! String
+        let sensorName = sensorDict["sensor_name"] as! String
+        let metaDict = sensorDict["meta"] as! Dictionary<String, AnyObject>
+        var sensorConfig = SensorConfig()
+        sensorConfig.meta = metaDict
         
-//closure 语法
-//        (parameters) -> returnType in
-//        
-//        statements
+        return Sensor(name: sensorName, source: sourceName, sensorConfig: sensorConfig, userId: KeychainWrapper.stringForKey(KEYCHAIN_USERID)!, remoteDataPointsDownloaded: false)
+    }
+    
+    func insertSensorDataIntoLocalDB(sensorData: JSON, sensorId: Int) throws {
```

Comment:
should these methods (one below as well) be private if they are helpers? They could also be un a util since they do not use state.